### PR TITLE
PyInstrument Extension

### PR DESCRIPTION
**Pull request by Helithumper**
## Description

This adds a new extension for PyInstrument, an instrumentation framework with HTML reports.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/1308

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Peyton Duncan for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: pyproject.toml [hunk 1]
```diff
@@ -61,6 +61,7 @@ channels = {version = ">=3.0.5", optional = true}
 astunparse = {version = "^1.6.3", python = "<3.9"}
 libcst = {version = ">=0.4.7", optional = true}
 rich = {version = ">=12.0.0", optional = true}
+pyinstrument = "^4.4.0"
```
Comment (patrick91) [inline pyproject.toml:64]: let's make this optional 😊

If I remember correctly we also need to create a group (like we did for django, fastapi)

File: strawberry/extensions/pyinstrument.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        profiler = Profiler()
+        profiler.start()
+
+        # Perform the operation
+        yield
+
+        # Stop the profiler
+        profiler.stop()
+        with Path.open(self._report_path, "w", encoding="utf-8") as f:
+            f.write(profiler.output_html())
```
Comment (patrick91) [inline strawberry/extensions/pyinstrument.py:38]: I think we can do `Path(self._report_path, "w", encoding="utf-8").write_text(...)`

File: tests/extensions/test_pyinstrument.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    )
+
+    # Query the schema
+    result = schema.execute_sync("{ theField }")
+
+    assert not result.errors
+    assert result.data["theField"] == 4
+
+    with Path.open(REPORT_PATH, encoding="utf-8") as f:
+        content = f.read()
```
Comment (patrick91) [inline tests/extensions/test_pyinstrument.py:45]: similar to above, `Path(REPORT_PATH, encoding="utf-8").read()`

File: RELEASE.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Release type: minor
+
+Added an extension for [PyInstrument](https://github.com/joerick/pyinstrument)
```
Comment (patrick91) [inline RELEASE.md:3]:
````suggestion
This releases adds an extension for [PyInstrument](https://github.com/joerick/pyinstrument). It allows to instrument your server and find slow code paths.

You can use it like this:

```python
import strawberry
from strawberry.extensions import pyinstrument
schema = strawberry.Schema(
    Query,
    extensions=[
        pyinstrument.PyInstrument(report_path="pyinstrument.html"),
    ],
)
```
````

**Review by patrick91 — COMMENTED** [review]
This looks great! I've added a few minor comments 😊

File: pyproject.toml (hunk 1 above)
Comment (Helithumper) [reply pyproject.toml:64]: Didn't know about optional dependencies, will check it out and modify it

File: strawberry/extensions/pyinstrument.py (hunk 2 above)
Comment (Helithumper) [reply strawberry/extensions/pyinstrument.py:38]: Another TIL. Will modify

Comment (Helithumper) [reply strawberry/extensions/pyinstrument.py:38]: Looking into this the other option is to perform the whole operation without the `with` context however I'm unsure if that's a better option as it just turns these two lines into one larger line without the context manager. Is there a specific reason we may want to do this in one line?

File: pyproject.toml (hunk 1 above)
Comment (Helithumper) [reply pyproject.toml]: Added, along with the group

File: strawberry/extensions/pyinstrument.py (hunk 2 above)
Comment (patrick91) [reply strawberry/extensions/pyinstrument.py:38]: mostly to remove the `with` statement, since it is done inside `write_text` 😊

Comment (patrick91) [reply strawberry/extensions/pyinstrument.py:38]: ah, you can also remove `w`: https://github.com/python/cpython/blob/3.11/Lib/pathlib.py#L1070-L1079

Comment (Helithumper) [reply strawberry/extensions/pyinstrument.py]: https://github.com/strawberry-graphql/strawberry/pull/2727/commits/b43ab659b032cfae5f1a413602fb2799632c2c67

File: tests/extensions/test_pyinstrument.py (hunk 3 above)
Comment (Helithumper) [reply tests/extensions/test_pyinstrument.py]: https://github.com/strawberry-graphql/strawberry/pull/2727/commits/b43ab659b032cfae5f1a413602fb2799632c2c67

**Review by patrick91 — APPROVED** [review]
Thank you so much for doing this :D

I've made a couple of fixes, but nothing major 😊

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Comment by Helithumper** [comment]
Thanks for finishing this up :D

**Pull request merged**